### PR TITLE
Implement top level Privacy Budget Manager

### PR DESCRIPTION
**Pull request by uakyol**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)

### CODE REVIEW

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @uakyol)

___
*[src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt, line 73 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HQuUN-A_R7C7jzftv:-N0HQuUN-A_R7C7jzftw:b-cgdsbw) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt#L73)):*
> ```Kotlin
>     val otherVidSampleEnd = otherBucketGroup.vidSampleStart + otherBucketGroup.vidSampleWidth
> 
>     // Vid ranges are non inclusive. vid(0.1, 0.2) does not overlap with vid(0.2, 0.3)
> ```

Is neither end of the range closed? e.g. `[0.1, 0.2)` as opposed to `(0.1, 0.2)`?



_Code quote:_
```Kotlin
non inclusive
```
___
*[src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt, line 145 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HRZTB8l3SGQHi908S:-N0HRZTB8l3SGQHi908T:b-pybjmz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L145)):*
> ```Kotlin
> 
>   @Test
>   fun `Privacy budget manager works as expected for reach and frequency measurement`() {
> ```

I assume the PBM will have more methods later, so these should probably be phrased as what the method does.

e.g. "chargePrivacyBudget throws PRIVACY_BUDGET_EXCEEDED when ..."

"works as expected" is overly vague, as it doesn't tell you what it's testing. It's also especially misleading for a test that verifies that something throws an error.



_Code quote:_
```Kotlin
Privacy budget manager works as expected for reach and frequency measurement
```
___
*[src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt, line 199 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HSDsM71E0bKkjV4Wj:-N0HSDsM71E0bKkjV4Wk:b-gu079e) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L199)):*
> ```Kotlin
>     assertThat(exception.errorType)
>       .isEqualTo(PrivacyBudgetManagerExceptionType.PRIVACY_BUDGET_EXCEEDED)
>   }
> ```

I don't see any test for cases where the charge is successful?

**Review by uakyol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: 3 of 5 files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @matthewclegg, and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt` line 73 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HQuUN-A_R7C7jzftv:-N0X7r6p7gHqemzRmAUh:b-4om15p) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt#L73)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Is neither end of the range closed? e.g. `[0.1, 0.2)` as opposed to `(0.1, 0.2)`?
</blockquote></details>

Thank you for catching. You are right, they should be half-open intervals. [0.1, 0.2) . Changed that. 
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 145 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HRZTB8l3SGQHi908S:-N0XJ4aDEiXIa6672O-n:badrzzj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L145)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I assume the PBM will have more methods later, so these should probably be phrased as what the method does.

e.g. "chargePrivacyBudget throws PRIVACY_BUDGET_EXCEEDED when ..."

"works as expected" is overly vague, as it doesn't tell you what it's testing. It's also especially misleading for a test that verifies that something throws an error.
</blockquote></details>

ah thanks. 
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 199 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HSDsM71E0bKkjV4Wj:-N0XHsh9Cp18CYcYrtH9:b-jaqoz6) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L199)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I don't see any test for cases where the charge is successful?
</blockquote></details>

so this one for example is the case where charge is successful. I borrowed this pattern from here : https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/1ebef8d84a0a0cde886b9525a02c2eff35504063/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetLedgerTest.kt#L60 

Basically, I make a charge above that works but fills the privacy budget and then make another charge that exceeds the privacy budget. Currently chargePrivacyBudget doesn't return anything. If you think it is a good idea, I can make it return a PrivacyBudgetReturnStatus. In fact that's how @matthewclegg implemented first but he changed it based on feedback. If you think we should do it like that, what do you think is the best way to handle budget exceeded: 

1) Throw Error(budget exceeded) from chargePrivacyBudget 
OR
2) Return PrivacyBudgetManagerReturnStatus(budget exceeded) from chargePrivacyBudget

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @matthewclegg, and @uakyol)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt` line 73 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HQuUN-A_R7C7jzftv:-N0XbjlFABJA1Du3jkwX:bs2ey8f) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt#L73)):*
<details><summary><i>Previously, uakyol wrote…</i></summary><blockquote>

Thank you for catching. You are right, they should be half-open intervals. [0.1, 0.2) . Changed that. 
</blockquote></details>

Close. You have `non half-open intervals`
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 145 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HRZTB8l3SGQHi908S:-N0Xc30nDFn4TGLxCGDL:b-y7eodt) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L145)):*
<details><summary><i>Previously, uakyol wrote…</i></summary><blockquote>

ah thanks. 
</blockquote></details>

You still have "works as expected" for a lot of these tests. I assume they should all be swapped to e.g. "throws PRIVACY_BUDGET_EXCEEDED when ..." or "charges privacy budget when/for ..."
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 199 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HSDsM71E0bKkjV4Wj:-N0XcQ4hDlSm0PFD5-Nx:b-pj1d6t) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L199)):*
<details><summary><i>Previously, uakyol wrote…</i></summary><blockquote>

so this one for example is the case where charge is successful. I borrowed this pattern from here : https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/1ebef8d84a0a0cde886b9525a02c2eff35504063/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetLedgerTest.kt#L60 

Basically, I make a charge above that works but fills the privacy budget and then make another charge that exceeds the privacy budget. Currently chargePrivacyBudget doesn't return anything. If you think it is a good idea, I can make it return a PrivacyBudgetReturnStatus. In fact that's how @matthewclegg implemented first but he changed it based on feedback. If you think we should do it like that, what do you think is the best way to handle budget exceeded: 

1) Throw Error(budget exceeded) from chargePrivacyBudget 
OR
2) Return PrivacyBudgetManagerReturnStatus(budget exceeded) from chargePrivacyBudget
</blockquote></details>

So what we want is to verify the effect of chargePrivacyBudget. Checking the side effect by seeing if the budget is exceeded by a second call is okay, as long as it's clear. The way the method is written, the first chargePrivacyBudget call looks like it's part of the setup block and the second call looks like it's what's being tested.

You also have access to the underlying backing store here, so that's another possible way to verify the effect of the function.

**Review by uakyol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: 3 of 5 files reviewed, 3 unresolved discussions (waiting on @duliomatos1, @matthewclegg, and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt` line 73 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HQuUN-A_R7C7jzftv:-N0XgFE7BVGhYcEs2p0Z:bw7uug7) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt#L73)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Close. You have `non half-open intervals`
</blockquote></details>

:man_facepalming:  thank you! 
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 145 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HRZTB8l3SGQHi908S:-N0Y_bup2MlbeHKCZKOW:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L145)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

You still have "works as expected" for a lot of these tests. I assume they should all be swapped to e.g. "throws PRIVACY_BUDGET_EXCEEDED when ..." or "charges privacy budget when/for ..."
</blockquote></details>

Done.
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 199 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0HSDsM71E0bKkjV4Wj:-N0Y_bJeD-0rDSVtQqb-:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c70f55773bb5e8bfb1b4b54ebbbb2c8a39606735/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L199)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

So what we want is to verify the effect of chargePrivacyBudget. Checking the side effect by seeing if the budget is exceeded by a second call is okay, as long as it's clear. The way the method is written, the first chargePrivacyBudget call looks like it's part of the setup block and the second call looks like it's what's being tested.

You also have access to the underlying backing store here, so that's another possible way to verify the effect of the function.
</blockquote></details>

Done.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @uakyol)

___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt` line 172 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0aqSg04oSM5BOdsiJ2:-N0aqSg04oSM5BOdsiJ3:b5qoxpf) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManagerTest.kt#L172)):*
> ```Kotlin
> 
>   @Test
>   fun `charges privacy budget for imprression measurement`() {
> ```

nit: impression



_Code quote:_
```Kotlin
imprression
```

**Review by duliomatos1 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: 4 of 5 files reviewed, 2 unresolved discussions (waiting on @SanjayVas and @uakyol)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt` line 74 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bD4MX01W-ZyhMY_eJ:-N0bD4MX01W-ZyhMY_eK:bfekp7t) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt#L74)):*
> ```Kotlin
> 
>     // Vid ranges are half-open intervals. [0.1, 0.2) does not overlap with vid[0.2, 0.3)
>     if (otherVidSampleEnd <= this.vidSampleStart ||
> ```

could simplify, this is equivalent in boolean algebra:

```
return this.vidSampleStart < otherVidSampleEnd && otherBucketGroup.vidSampleStart < thisVidSampleEnd
```
___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManager.kt` line 74 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bBZYB5gHO82tHF1Fs:-N0bBZYB5gHO82tHF1Ft:b-w18gow) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManager.kt#L74)):*
> ```Kotlin
>           )
>         )
>       // TODO(@uakyol): After the privacy budget accounting is switched to using the Gaussian
> ```

you could extract the current method and new method as functions if you think it would look better for now
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroupTest.kt` line 83 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bAheH-XBoECWF9vU8:-N0bAheH-XBoECWF9vU9:btv8qh4) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroupTest.kt#L83)):*
> ```Kotlin
> 
>   @Test
>   fun `overlapsWith non overlapping works as expected for neighbor bucket groups`() {
> ```

nit: maybe something like this to avoid phrasing with "works as expected" 
`overlapsWith() returns false for two neighboring, non-overlapping, bucket groups`

**Review by uakyol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539)* status: 4 of 5 files reviewed, 2 unresolved discussions (waiting on @duliomatos1 and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt` line 74 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bD4MX01W-ZyhMY_eJ:-N0bJMJr2H2sfKvkFfXA:b-vambkj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroup.kt#L74)):*
> return this.vidSampleStart < otherVidSampleEnd && otherBucketGroup.vidSampleStart < thisVidSampleEnd
ah thanks, sounds good.
___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManager.kt` line 74 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bBZYB5gHO82tHF1Fs:-N0bJsHU8N-vkoqgmi7A:b-ha80rq) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManager.kt#L74)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

you could extract the current method and new method as functions if you think it would look better for now
</blockquote></details>

Lets leave it like this if you don't mind. The old way of doing this will be deleted anyway,.
___
*[`src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroupTest.kt` line 83 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bAheH-XBoECWF9vU8:-N0bJcry59YZ33zQwQ_F:b-ayo7a2) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/test/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBucketGroupTest.kt#L83)):*
> overlapsWith() returns false for two neighboring, non-overlapping, bucket groups
much better, thank you

**Review by duliomatos1 — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 3 files at r5, 2 of 5 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-:-N0blkrP1Xccart9KPEV:bg7igkv)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManager.kt` line 74 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/539#-N0bBZYB5gHO82tHF1Fs:-N0blbkt9r0BtDRGnkqa:bfwf456) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/e1834d54d8fca787e1862cbf607924c98b42b2d9/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/PrivacyBudgetManager.kt#L74)):*
<details><summary><i>Previously, uakyol wrote…</i></summary><blockquote>

Lets leave it like this if you don't mind. The old way of doing this will be deleted anyway,.
</blockquote></details>

I'm good with that.
